test/metrics/prom-metrics-linter/custom_linter_rules.go

```diff
@@ -56,10 +57,161 @@ func CustomLinterRules(problems []promlint.Problem, mf *dto.MetricFamily, operat
+		})
+	}
+	return problems
+}
+
+func validateRecordingRuleOpsSuffix(rr recordingRule, parts []string) []promlint.Problem {
+	name := rr.Record
+	var problems []promlint.Problem
+
+	aggOps := detectOps(rr.Expr, promAggOps)
+	timeOps := detectOps(rr.Expr, promTimeOps)
+	total := len(aggOps) + len(timeOps)
+	opsPart := parts[2]
+
+	if total == 1 {
+		if len(aggOps) == 1 {
+			expected := aggOps[0]
+			if !strings.HasSuffix(opsPart, expected) {
+				problems = append(problems, promlint.Problem{
+					Metric: name,
+					Text:   fmt.Sprintf("single aggregation detected in expr: operations (third segment in level:metric:operations) should end with %q", expected),
+				})
+			}
+		} else if len(timeOps) == 1 {
+			expected := timeOps[0]
+			timeSuffixRe, err := regexp.Compile(`(?i)` + expected + `[0-9]+[smhdwy]+$`)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "time op suffix regex compile failed for %q: %v\n", expected, err)
+			} else if !timeSuffixRe.MatchString(opsPart) {
+				problems = append(problems, promlint.Problem{
+					Metric: name,
+					Text:   fmt.Sprintf("single time operation detected in expr: operations (third segment in level:metric:operations) should end with %q", expected+"<duration>"),
+				})
+			}
+		}
+	} else if total > 1 {
+		present := false
+		for _, op := range append(aggOps, timeOps...) {
+			if strings.Contains(opsPart, op) {
+				present = true
+				break
+			}
+		}
+		if !present {
+			problems = append(problems, promlint.Problem{
+				Metric: name,
+				Text:   "multiple operations detected in expr: at least one operation should appear in operations (third segment in level:metric:operations)",
+			})
+		}
+	}
+	return problems
+}
+
+func validateRecordingRuleNoDuplicateOps(name string, parts []string) []promlint.Problem {
+	var problems []promlint.Problem
+	seen := map[string]struct{}{}
+	tokens := strings.Split(parts[2], "_")
+	for _, tok := range tokens {
+		if tok == "" {
+			continue
+		}
+		if _, ok := seen[tok]; ok {
+			problems = append(problems, promlint.Problem{
+				Metric: name,
+				Text:   "operations (third segment in level:metric:operations) contains duplicate tokens; merge duplicates (e.g., min_min -> min)",
+			})
+			break
+		}
+		seen[tok] = struct{}{}
+	}
+	return problems
+}
+
+var (
+	promAggOps = []string{
+		// Aggregations
+		"sum", "avg", "min", "max", "count", "quantile", "stddev", "stdvar",
+		"group", "count_values", "limitk", "limit_ratio",
+		// Selectors considered as aggregations
+		"topk", "bottomk",
+	}
+	promTimeOps = []string{
+		// Time/transform functions
+		"rate", "irate", "increase", "delta", "idelta", "deriv",
+		"avg_over_time", "min_over_time", "max_over_time", "sum_over_time",
+		"count_over_time", "quantile_over_time", "stddev_over_time", "stdvar_over_time",
+	}
+)
+
+func detectOps(expr string, ops []string) []string {
+	pattern := "\\b(" + strings.Join(ops, "|") + ")\\b"
+	re, err := regexp.Compile(pattern)
```

Comment:
This is based const lists of strings, so we can compile in advance, only once. Compiling a regex is pricey, and we should not do it in loops, for the same input.